feat(python): allow construction of `Datetime` series from `datetime.date` array

mcrumiller opened the pull request:
Resolves #12150

MarcoGorelli commented on py-polars/tests/unit/series/test_series.py:
```diff
@@ -117,6 +117,13 @@ def test_init_inputs(monkeypatch: Any) -> None:
     # lists
     assert pl.Series("a", [[1, 2], [3, 4]]).dtype == pl.List(pl.Int64)
 
+    # conversion of Date to Datetime
+    s = pl.Series([date(2023, 1, 1), date(2023, 1, 2)], dtype=pl.Datetime)
+    assert s.to_list() == [datetime(2023, 1, 1), datetime(2023, 1, 2)]
+    assert Datetime == s.dtype
+    assert s.dtype.time_unit == "us"  # type: ignore[union-attr]
```
what happens if someone does
```python
s = pl.Series([date(2023, 1, 1), date(2023, 1, 2)], dtype=pl.Datetime('ms'))
```
?

mcrumiller replied:
Good point.

mcrumiller replied:
Updated, let me know how that looks.

MarcoGorelli commented on py-polars/polars/utils/_construction.py:
```diff
@@ -491,35 +489,39 @@ def sequence_to_pyseries(
-                        " To convert to a different time zone, please use `.dt.convert_time_zone`."
+            if dtype == Datetime:
+                if values_dtype == Date:
+                    # check for zone info
+                    return (
+                        s.cast(Datetime(time_unit)).dt.replace_time_zone(time_zone)._s
                     )
```
is there a way to do this as an earlier return, rather than adding an extra level of indentation (to already quite indented code)?

mcrumiller replied:
I could remove one level of indendation by moving it just a bit earlier with

```python
if (values_dtype == Date) and (dtype == Datetime):
    return s.cast(Datetime(time_unit)).dt.replace_time_zone(time_zone)._s
```

Which would save us 2 lines since the formatter will remove the parens. Let's try it.

MarcoGorelli commented on py-polars/tests/unit/series/test_series.py:
```diff
@@ -117,6 +121,24 @@ def test_init_inputs(monkeypatch: Any) -> None:
+    assert s.dtype.time_zone is None  # type: ignore[union-attr]
+
+    # conversion of Date to Datetime with specified timezone and units
+    tu: Final = "ms"
+    tz: Final = "America/Argentina/Rio_Gallegos"
```
final looks a bit odd here 😄 I think it's usually used in classes to indicate that things shouldn't be overridden, can we remove it?

mcrumiller replied:
The `Final` keyword is because the signature for `pl.Date`:

```python
def datetime_(
    ...
    time_unit: TimeUnit = "us",
    ...
) -> Expr:
```
includes the `TimeUnit` type alias which is defined as  `TimeUnit: TypeAlias = Literal["ns", "us", "ms"]`. MyPy doesn't accept strings as literals unless they are labeled as `Final`.

I can remove it from the `tz`, however, or I can put a `# type:ignore` in there if you want to clean up the `Final` stuff.

MarcoGorelli replied:
Does it work to annotate it as `TimeUnit` instead?

mcrumiller replied:
Worked like a charm! Thank you.

MarcoGorelli commented on py-polars/polars/utils/_construction.py:
```diff
@@ -491,35 +489,39 @@ def sequence_to_pyseries(
-                        " To convert to a different time zone, please use `.dt.convert_time_zone`."
+            if dtype == Datetime:
+                if values_dtype == Date:
+                    # check for zone info
```
I'm confused by this comment

mcrumiller replied:
Me too, removing.

MarcoGorelli reviewed (commented):
I think the logic's fine, just got some minor comments

mcrumiller commented:
@stinodego thanks for the suggestions. I cleaned a few things up:

* Made a special constructor case for datetime w/ dates, which resulted in a few lines saved and the whole datetime block de-nested one level.
* Removed confusing comment.
* Removed one unnecessary `Final`. The other is still necessary since the datetime constructor expects a `Literal` instead of a string. If you want to remove the import/Final keywords we can put a mypy ignore in there instead.

MarcoGorelli commented:
Thanks @mcrumiller
